adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -940,93 +947,110 @@ protected void setNonEmptyResponsePayload(final HttpServerResponse response, fin
+                            tenantObject.getTenantId(),
+                            tenantObject,
+                            ProcessingOutcome.UNDELIVERABLE,
+                            command.getPayloadSize(),
+                            commandSample);
+                    log.debug("command for device has already fired [tenantId: {}, deviceId: {}]",
+                            tenantObject.getTenantId(), deviceId);
+                    commandContext.release();
+                } else {
+                    checkMessageLimit(tenantObject, command.getPayloadSize())
+                            .setHandler(result -> {
+                                if (result.succeeded()) {
+                                    addMicrometerSample(commandContext, commandSample);
+                                    // put command context to routing context and notify
+                                    ctx.put(CommandContext.KEY_COMMAND_CONTEXT, commandContext);
+                                } else {
+                                    // issue credit so that application(s) can send the next command
+                                    commandContext.reject(getErrorCondition(result.cause()), 1);
+                                    metrics.reportCommand(
+                                            command.isOneWay() ? Direction.ONE_WAY : Direction.REQUEST,
+                                            tenantObject.getTenantId(),
+                                            tenantObject,
+                                            ProcessingOutcome.from(result.cause()),
+                                            command.getPayloadSize(),
+                                            commandSample);
+                                }
+                                cancelCommandReceptionTimer(ctx);
+                                setTtdStatus(ctx, TtdStatus.COMMAND);
+                                responseReady.tryComplete();
+                            });
+                }
+            } else {
+                getMetrics().reportCommand(
+                        command.isOneWay() ? Direction.ONE_WAY : Direction.REQUEST,
+                        tenantObject.getTenantId(),
+                        tenantObject,
+                        ProcessingOutcome.UNPROCESSABLE,
+                        command.getPayloadSize(),
+                        commandSample);
+                log.debug("command message is invalid: {}", command);
+                commandContext.reject(new ErrorCondition(Constants.AMQP_BAD_REQUEST, "malformed command message"));
+            }
+            // we do not issue any new credit because the
+            // consumer is supposed to deliver a single command
+            // only per HTTP request
+        };
+
+        final Handler<Void> remoteCloseHandler = remoteDetach -> {
+            log.debug("peer closed command receiver link [tenant-id: {}, device-id: {}]", tenantObject.getTenantId(), deviceId);
```

Comment:
what about including the gateway ID here as well?

Reply:
Yes, I've added that.